stackslib/src/chainstate/burn/operations/leader_block_commit.rs

```diff
@@ -595,110 +641,155 @@ impl LeaderBlockCommitOp {
-                            error!("Failed to check whether parent (height={}) is descendent of anchor block={}: {}",
-                                   parent_block_height, &reward_set_info.anchor_block, e);
-                            op_error::BlockCommitAnchorCheck})?;
-                    if descended_from_anchor != expect_pox_descendant {
-                        if descended_from_anchor {
-                            warn!("Invalid block commit: descended from PoX anchor {}, but used burn outputs", &reward_set_info.anchor_block);
-                        } else {
-                            warn!("Invalid block commit: not descended from PoX anchor {}, but used PoX outputs", &reward_set_info.anchor_block);
-                        }
+            // we've checked length equality, so we can just iterate through
+            //  self.commit_outs and check if each is in `check_recipients`
+            //  *OR* if `allows_pox_punishment`, then it could be a burn.
+            // NOTE: we do a find and remove here so that the same recipient
+            //  isn't found multiple times by different commit_outs.
+            let mut rewarded = vec![];
+            for self_commit in self.commit_outs.iter() {
+                let search_predicate = self_commit.to_burnchain_repr();
+                let found = check_recipients
+                    .iter()
+                    .enumerate()
+                    .find(|(_, (check_commit, _))| {
+                        search_predicate == check_commit.to_burnchain_repr()
+                    });
+                if let Some((index, _)) = found {
+                    rewarded.push(Treatment::Reward(check_recipients.remove(index).0));
+                } else {
+                    // if we didn't find the pox output, then maybe its a pox punishment?
+                    if reward_set_info.allow_nakamoto_punishment && self_commit.is_burn() {
+                        continue;
+                    } else {
+                        warn!("Invalid block commit: committed output {} does not match expected recipient set: {:?}",
+                              self_commit.to_burnchain_repr(), check_recipients);
                         return Err(op_error::BlockCommitBadOutputs);
                     }
-                }
+                };
             }
-        } else {
-            // no recipient info for this sortition, so expect all burns
-            if !self.all_outputs_burn() {
-                warn!("Invalid block commit: this transaction should only have burn outputs.");
+
+            if !descended_from_anchor {
+                warn!(
+                    "Invalid block commit: not descended from PoX anchor {}, but used PoX outputs",
+                    &reward_set_info.anchor_block
+                );
                 return Err(op_error::BlockCommitBadOutputs);
             }
-        };
-        Ok(())
+
+            let mut treated_outputs: Vec<_> = check_recipients
```

Comment:
I'm trying to understand something here.  Is it possible that the commit-outs PoX address ordering in the block-commit gets changed by this?  It seems that if there's one address that's up for punishment in a reward phase, then it will always be the _first_ PoX address in `treaded_outputs`.  Is that desirable, and if so, why is it more desirable than keeping the addresses in `treaded_outputs` in the same order as they are in `commit_outs`?

Reply:
No `.commit_outs` can't be reordered because its only accessed through the non-mutating `iter()`.

If you're asking about the ordering of `treated_outputs`, then yes, the ordering of `treated_ouputs` can be different than the ordering of `.commit_outs` (and both may be different than the ordering of `check_recipients`. I don't think the ordering of `treated_outputs` matters.

Reply:
I was asking about the latter.  Just wanted to make sure the test coverage indicated that the ordering of `treated_outputs` does not matter.